src/Servers/IIS/IIS/test/Common.FunctionalTests/AppOfflineTests.cs

```diff
@@ -131,17 +132,64 @@ public async Task AppOfflineDroppedWhileSiteStarting_SiteShutsDown_InProcess()
                         // if AssertAppOffline succeeded ANCM have picked up app_offline before starting the app
                         // try again
                         RemoveAppOffline(deploymentResult.ContentRoot);
+
+                        if (deploymentResult.DeploymentParameters.ServerType == ServerType.IIS)
+                        {
+                            deploymentResult.AssertWorkerProcessStop();
+                            return;
+                        }
                     }
                     catch
                     {
+                        // For IISExpress, we need to catch the exception because IISExpress will not restart a process if it crashed.
+                        // RemoveAppOffline will fail due to a bad request exception as the server is down.
+                        Assert.Contains(TestSink.Writes, context => context.Message.Contains("Drained all requests, notifying managed."));
                         deploymentResult.AssertWorkerProcessStop();
                         return;
                     }
                 }
 
                 Assert.True(false);
+            }
+        }
+
+        [ConditionalFact]
+        public async Task GracefulShutdownWorksWithMultipleRequestsInFlight_InProcess()
+        {
+            var deploymentParameters = _fixture.GetBaseDeploymentParameters(_fixture.InProcessTestSite);
+            deploymentParameters.TransformArguments((a, _) => $"{a} IncreaseShutdownLimit");
+
+            var deploymentResult = await DeployAsync(deploymentParameters);
+
+            var result = await deploymentResult.HttpClient.GetAsync("/HelloWorld");
+
+            var connectionList = new List<TestConnection>();
+
+            for (var i = 0; i < 3; i++)
+            {
+                var connection = new TestConnection(deploymentResult.HttpClient.BaseAddress.Port);
+                await connection.Send(
+                    "POST /HelloWorldDelayed HTTP/1.1",
+                    "Content-Length: 1",
+                    "Host: localhost",
+                    "Connection: close",
+                    "",
+                    "");
+
+                connectionList.Add(connection);
+            }
+
+            // Shouldn't stop the server here. Instead should drop app offline 
+            AddAppOffline(deploymentResult.ContentRoot);
```

Comment:
I wonder if the server would even be able to detect and react to app_offline before we get to `connectionList[i].Send/Receive`

Reply:
Good point.